C++ client producer sendAsync() method will be blocked forever, if enough batched messages sent timeout.

**Describe the bug**
C++ client producer sendAsync() method will be blocked forever, if enough batched messages sent timeout.
**To Reproduce**

1. Using the following producer configures: 
setBatchingEnabled(true)
setBlockIfQueueFull(true)
setSendTimeout(100) (shorter timeout interval would help reproduce the issue)
2. Producing huge plant of messages (better in a pressure test env.);
3. Try to made some producer timeout happens (with shorter timeout interval for example)
4. After accumulating enough timeout captured, the pulsar::Producer::sendAsync method will be blocked and will nerver turned back to available.

**Expected behavior**
The method pulsar::Producer::sendAsync(...) might be blocked for a while, if the pending messages reached the max value, but it shall be returned back available if previous pending messages acked or reaching timeout point.


**Desktop (please complete the following information):**
 - OS: Centos

A weird Unit test issue encountered...
======================================================================
FAIL: test_publish_compact_and_consume (__main__.PulsarTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pulsar_test.py", line 641, in test_publish_compact_and_consume
    self.assertEqual(msg0.data(), b'hello-1')
AssertionError: 'hello-0' != 'hello-1'

----------------------------------------------------------------------

From the code, seems the case is something about the consumer configuration: "reading compact".
Very strange.

--- a/pulsar-client-cpp/lib/ProducerImpl.cc
+++ b/pulsar-client-cpp/lib/ProducerImpl.cc
@@ -232,6 +232,14 @@ void ProducerImpl::failPendingMessages(Result result) {
     // without holding producer mutex.
     for (MessageQueue::const_iterator it = pendingMessagesQueue_.begin(); it != pendingMessagesQueue_.end();
          it++) {
+        // When dealing any failure message, if the current message is a batch one, we should also release
+        // the reserved spots in the pendingMessageQueue_, for all individual messages inside this batch
+        // message. See 'ProducerImpl::sendAsync' for more details.
+        if (it->msg_.impl_->metadata.has_num_messages_in_batch()) {
+            // batch message - need to release more spots
+            // -1 since the pushing batch message into the queue already released a spot
+            pendingMessagesQueue_.release(it->msg_.impl_->metadata.num_messages_in_batch() - 1);
+        }
         messagesToFail.push_back(*it);
     }
 

